Support Dark Mode

cselin opened the pull request:
@nickoneill I did a quick spike on the status of dark mode for SwiftUI and it's looking pretty good. I made a few quick changes to get something decent, but we probably want to add official dark appearance colors for custom colors. Let me know what you think!

|About|My Impact|
|-|-|
|![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-11-03 at 13 58 07](https://github.com/5calls/ios/assets/810263/06d84179-31a7-4c42-b54f-a2de05035c90)|![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-11-03 at 13 58 03](https://github.com/5calls/ios/assets/810263/e37bd08b-787d-4622-af97-2872536d4045)|

|Reminders|Location|
|-|-|
|![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-11-03 at 13 57 58](https://github.com/5calls/ios/assets/810263/c5ec040f-ba55-4184-b26e-3c983450ea8c)|![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-11-03 at 13 57 52](https://github.com/5calls/ios/assets/810263/0c92a523-701f-4edd-8e20-a29806b3cb67)|

|Detail|Script|
|-|-|
|![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-11-03 at 13 57 34](https://github.com/5calls/ios/assets/810263/c13bde79-20b8-4232-b853-2d8104784673)|![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-11-03 at 13 57 42](https://github.com/5calls/ios/assets/810263/7c313e77-72a8-4476-bda8-f704ed8265c0)|

cselin commented on FiveCalls/FiveCalls/ScheduleReminders.swift:
```diff
@@ -158,8 +158,6 @@ struct DayAndTimePickers: View {
             DatePicker("", selection: $selectedTime, displayedComponents: .hourAndMinute)
                 .labelsHidden()
                 .datePickerStyle(WheelDatePickerStyle())
-                .colorInvert() // TODO: this won't work in dark mode
```
I can't seem to find a way to change the color of the date picker wheel text that's compatible with dark mode unfortunately so we might have to use defaults here.

nickoneill replied:
👍 it looks OK as-is

nickoneill commented:
I mentioned to @heathermh the other day that I wasn't sure dark mode fits with the brand overall, but I don't dislike it after seeing the real thing. The text colors definitely need some customization for the dark background but very nice overall!

`PrimaryButton` looks fine, `fivecallsDarkBlueText` probably needs to be brighter on dark mode. We could remove the colored text on both Impact and Reminders in favor of size / weight changes to accentuate the text hierarchy.

heathermh commented:
@cselin Dark mode is definitely something I'm excited to have. This is looking good so far.

cselin commented:
@nickoneill I'm going to pick this up again this week, now that you've deleted the UIKit code

cselin commented:
@heathermh I ended up reverting the changes for red in dark mode to match light mode as I think it looks fine as text and definitely better for the button, that is unless with think coloring the donate button red is a bit misleading as it is often used for warnings or destructive actions:

|Before|After|
|-|-|
|![Simulator Screenshot - iPhone 15 - 2023-12-04 at 13 43 11](https://github.com/5calls/ios/assets/810263/6d359d02-db39-4563-9706-7fb60a503a46)|![Simulator Screenshot - iPhone 15 - 2023-12-04 at 13 45 15](https://github.com/5calls/ios/assets/810263/a256a6db-c4d8-4a85-9f0a-bd5e473add24)|

@nickoneill take a look at how Your Impact looks without the colors. I think the colors actually add something here in both light and dark mode, but we can remove:

|Light Before|Light After|
|-|-|
|![Simulator Screenshot - iPhone 15 - 2023-12-04 at 13 53 43](https://github.com/5calls/ios/assets/810263/90d97316-4f80-4f85-94a9-fececf57b6ea)|![Simulator Screenshot - iPhone 15 - 2023-12-04 at 13 53 28](https://github.com/5calls/ios/assets/810263/e9173814-dec8-4385-86b3-6bfca5c2b872)|

|Dark Before|Dark After|
|-|-|
|![Simulator Screenshot - iPhone 15 - 2023-12-04 at 13 51 03](https://github.com/5calls/ios/assets/810263/40ce0b49-5eaf-4cb8-b3ac-5e70d56f3d96)|![Simulator Screenshot - iPhone 15 - 2023-12-04 at 13 53 19](https://github.com/5calls/ios/assets/810263/2e3b0343-4fcb-4452-9e68-be5f2da7ae3d)|

heathermh commented:
> @heathermh I ended up reverting the changes for red in dark mode to match light mode as I think it looks fine as text and definitely better for the button, that is unless with think coloring the donate button red is a bit misleading as it is often used for warnings or destructive actions:
I agree that the white text on that red button looks nice and gives a little interest to that page. My worry is when we use red text, especially when it's small. I ran into that on the reminders page where we select a time and days of the week. The contrast for text that small wasn't great. We could skip using red for text, maybe make the text bigger, or highlight it some other way - maybe a red symbol or something? Open to suggestions to keep the text readable. I use this contrast calculator since it's based on upcoming accessibility guidelines: https://www.myndex.com/APCA/. It suggests that fivecallsRed is almost OK for body text (17pt default) on a white background at regular weight. It's totally unusable on black.

nickoneill commented:
Agreed on changing the red for the donate button, the lighter color doesn't work on dark mode. Perhaps it deserves two variants like blue has, text color and element color?

Let's simply remove text colors on Impact for now. It's going to be hard to merge anything more significant with @heathermh's accessibility changes to that page.

cselin commented:
@heathermh @nickoneill updated with your suggestions. Added new `fivecallsRedText` color to use on Reminders with acceptable lighter red for dark mode, removed text colors from Your Impact and kept single `fivecallsRed` element color for donate button in light and dark.

heathermh commented:
> @heathermh @nickoneill updated with your suggestions. Added new `fivecallsRedText` color to use on Reminders with acceptable lighter red for dark mode, removed text colors from Your Impact and kept single `fivecallsRed` element color for donate button in light and dark.

I'm appreciative if somebody wants to try to improve the dark mode version of the red text color. All I did was lighten it up enough to create sufficient contrast / make it readable on black or very dark grey. It surely doesn't look very red, though.

nickoneill reviewed (approved):
Looks great 👍 I added a note in #331 to take a look at the launch screen in dark mode, not required to merge this though

cselin commented:
Updated `fivecallsRedText` to be a bit brighter and acceptable on APCA. Decided to apply it only to reminders grid and text since it made the Donate button way too bright on the done page. If this looks good, I'll merge these changes.
![Simulator Screenshot - iPhone 15 - 2023-12-13 at 11 03 21](https://github.com/5calls/ios/assets/810263/af2a4dc1-bb76-41d9-87f6-a21997f99544)

nickoneill commented:
👍 merge away

heathermh commented:
@cselin As much as I'd like to keep the red text, I'm worried that the bright red on very dark grey isn't enough contrast to make it readable. If I do a color picker from the screen shots below, I get red #EA4338 and grey #1C1C1E (APCA is -34 which is only OK for spot text). How about we keep the boxes red and the text white for dark mode? I've added some other accessibility bits for the reminders, but it's on the branch I'm working on. I'm happy to merge whatever is needed once I do a PR for the bit I'm working on.

cselin commented:
> @cselin As much as I'd like to keep the red text, I'm worried that the bright red on very dark grey isn't enough contrast to make it readable. If I do a color picker from the screen shots below, I get red #EA4338 and grey #1C1C1E (APCA is -34 which is only OK for spot text). How about we keep the boxes red and the text white for dark mode? I've added some other accessibility bits for the reminders, but it's on the branch I'm working on. I'm happy to merge whatever is needed once I do a PR for the bit I'm working on.

@heathermh feel free to make any additional changes to these as you work on reminders. Though wouldn't using red for the boxes have a similar issue with contrast as the text as the size of the lines is similar to the font size?

heathermh commented:
> @heathermh feel free to make any additional changes to these as you work on reminders. Though wouldn't using red for the boxes have a similar issue with contrast as the text as the size of the lines is similar to the font size?

Having to read regular size red text on a dark grey background is different than seeing lines in the same colors. We have to be able to see that text more clearly in order to make out the letters. If the font was bigger, that color combo might be readable.

Understanding the letters is higher priority than seeing / kind of seeing the lines. Users who can see the red, but not quite the red letters might still be able to get some functionality from the red lines. I do have some code on the branch I'm working on to add backup / supportive functionality for the red lines. If a user is using VoiceOver, differentiate without color or reduce animation, I have an alert that tells them they need to pick a day. That way if they don't see the red or don't want the shake, they get an alert to make sure they can understand what's going on.

cselin commented:
> > @heathermh feel free to make any additional changes to these as you work on reminders. Though wouldn't using red for the boxes have a similar issue with contrast as the text as the size of the lines is similar to the font size?
> 
> Having to read regular size red text on a dark grey background is different than seeing lines in the same colors. We have to be able to see that text more clearly in order to make out the letters. If the font was bigger, that color combo might be readable.
> 
> Understanding the letters is higher priority than seeing / kind of seeing the lines. Users who can see the red, but not quite the red letters might still be able to get some functionality from the red lines. I do have some code on the branch I'm working on to add backup / supportive functionality for the red lines. If a user is using VoiceOver, differentiate without color or reduce animation, I have an alert that tells them they need to pick a day. That way if they don't see the red or don't want the shake, they get an alert to make sure they can understand what's going on.

@heathermh makes sense, thanks for the details!

nickoneill commented:
@heathermh if there are other changes you think we should make here, can you open an issue about it? We'll forget about this one if it's only in the comments of a closed PR

heathermh commented:
@cselin you're welcome

@nickoneill I've got a few updates pending on the branch I'm working on plus I don't have a clear picture of where Main is at at this moment. After this branch I'll try to see where we're at and add issue(s) once I know if they're still there.
